smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/SmithyIntegration.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.codegen.core;
+
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.utils.SmithyUnstableApi;
+
+/**
+ * This interface provides the base concept of an "Integration" to
+ * Smithy code generators.
+ *
+ * <p>This class provides the bare minimum that most Smithy code generators
+ * can implement as a tool to customize generators. Code generators are
+ * expected to extend this interface to add various hooks to their generator
+ * (e.g., register protocol generators, register auth scheme integrations,
+ * attach middleware, intercept and augment CodeWriter sections, etc).
+ *
+ * <p>This interface is currently unstable as more requirements
+ * may be added in the future.
+ *
+ * @param <S> The settings object used to configure the generator.
+ */
+@SmithyUnstableApi
+public interface SmithyIntegration<S extends SmithyCodegenSettings> {
```

Comment:
Yeah this is an interesting problem. It would be nice to have a centralized solution around it that's flexible enough for any generator. Let me noodle on this.

Reply:
I took your idea and built on it a bit. You can now order integrations using a priority or a relative ordering based on before/after other integrations by name. I also included a topological sort code generators can use to order integrations.